master/buildbot/worker/upcloud.py

```diff
@@ -0,0 +1,167 @@
+
+    @defer.inlineCallbacks
+    def _resolve_image(self, image):
+        # get templates
+        result = yield self.client.get("/storage/template")
+        uuid = None
+        if (result.code == 200):
+            templates = yield result.json()
+            for template in templates["storages"]["storage"]:
+               if template["title"] == image:              
+                  uuid = template["uuid"]
+                  break
+        defer.returnValue(uuid)
+
+    @defer.inlineCallbacks
+    def start_instance(self, build):
+        if self.instance is not None:
+            raise ValueError('instance active')
+        # convert image to UUID
+        image = yield build.render(self.image)
+        image = yield self._resolve_image(image)
+        hostconfig = yield build.render(self.hostconfig)
+         # compose json
+        req = {
+            "server":{
+                "zone":hostconfig.get('zone', DEFAULT_ZONE),
+                "title":"Buildbot build #%d" % (self.substantiation_build.number,),
+                "hostname":hostconfig.get('hostname', self.name),
+                "user_data":hostconfig.get('user_data',""),
+                "login_user":{
+                    "username":"root",
+                    "ssh_keys": {
+                        "ssh_key": hostconfig.get('ssh_keys',[]),
+                    },
+                },
+                "storage_devices":{
+                    "storage_device":[{
+                        "action":"clone",
+                        "storage":image,
+                        "title":"Buildbot build #%d" % (self.substantiation_build.number,),
+                        "size":hostconfig.get("os_disk_size", DEFAULT_OS_DISK_SIZE),
+                        "tier":"maxiops",
+                    }],
+                }
+            }
+        }
+
+        req["server"]["plan"] = hostconfig.get("plan", DEFAULT_PLAN)
+        if req["server"]["plan"] == "custom":
+            req["server"]["core_number"] = hostconfig.get("core_number", DEFAULT_CORE_NUMBER)
+            req["server"]["memory_amount"] = hostconfig.get("memory_amount", DEFULT_MEMORY_AMOUNT)
+
+        # request instance
+        result = yield self.client.post("/server", json=req)
+
+        if int(result.code / 100) == 2:
+            instance = yield result.json()
+            self.instance = instance["server"]
+            self.instance["Id"] = self.instance["uuid"].split("-")[-1]
+            log.msg("UpcloudLatentWorker: Instance %s created (root password %s)" % (self.instance["Id"],self.instance['password']))
+            defer.returnValue([self.instance["Id"], image])
+        else:
+            self.failed_to_start(req['server']['hostname'], image)
+
+    @property
+    def shortid(self):
+        if self.instance is None:
+            return None
+        return self.instance['Id'][:6]
+
+    @defer.inlineCallbacks
+    def state(self):
```

Comment:
I guess this is a private method
```suggestion
    def _state(self):
```